extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -642,19 +676,54 @@ export class AmpAutocomplete extends AMP.BaseElement {
     if (delta === 0 || !this.resultsShowing_()) {
       return Promise.resolve();
     }
+    // Active element logic
     const keyUpWhenNoneActive = this.activeIndex_ === -1 && delta < 0;
     const index = keyUpWhenNoneActive ? delta : this.activeIndex_ + delta;
     const activeIndex = mod(index, this.container_.children.length);
     const newActiveElement = this.container_.children[activeIndex];
     this.inputElement_.value = newActiveElement.getAttribute('data-value');
-    return this.mutateElement(() => {
+
+    // Element visibility logic
+    let overflowEl;
+
+    return this.measureMutateElement(() => {
+      overflowEl = this.measureContainerScroll_(newActiveElement, delta > 0);
+    }, () => {
+      if (overflowEl.shouldScroll) {
+        this.container_.scrollTop = overflowEl.newTop;
+      }
       this.resetActiveElement_();
       newActiveElement.classList.add('i-amphtml-autocomplete-item-active');
       this.activeIndex_ = activeIndex;
       this.activeElement_ = newActiveElement;
     });
   }
 
+  /**
+   * Given an element and boolean direction, returns an object of the format:
+   * { shouldScroll: boolean, newTop: number }
+   *
+   * This method assumes that the element it is passed is a child
+   * of the results container, otherwise known as a suggested item.
+   * It then calculates whether the item element is visible within
+   * the scope of the container element, and if not, what the container
+   * should be scrolled to (either upwards or downwards),in order for the
+   * element to be seen. Should be called in a measureMutate context.
+   *
+   * @param {!Element} element
+   * @param {boolean} goingDown
+   * @return {!Object}
+   * @private
+   */
+  measureContainerScroll_(element, goingDown) {
+    const {offsetTop: itemTop, offsetHeight: itemHeight} = element;
+    const resultHeight = this.container_.offsetHeight;
+    const shouldScroll = (this.container_.scrollTop > itemTop ||
+      this.container_.scrollTop + resultHeight < itemTop + itemHeight);
+    const newTop = goingDown ? itemTop + itemHeight - resultHeight : itemTop;
+    return {shouldScroll, newTop};
```

Comment:
It's kind of unusual to return a tuple like this. On one hand, I see how you need to read the same values to know whether to change the scroll, but on the other hand returning tuples is a code smell. What if you break this out of a method and put it in the measure block? Since this method isn't reused, it should be OK not being a function to avoid having two return values. What do you think?

Reply:
That sounds reasonable to me. Another option could be to overload the return in a nullable number. That way, `null` means it shouldn't scroll, and not `null` is the value that should be scrolled to.

Reply:
That's another idea. It can be useful to give null a semantic value in some cases to communicate different states, but it can become a footgun if someone assumes the method always returns a primitive value and attempts to use the null value in a calculation. It's all about tradeoffs though, since that solves both the reusability problem and the tuple-return-value problem. Up to you!